perun-core/src/test/java/cz/metacentrum/perun/core/entry/GroupsManagerEntryIntegrationTest.java

```diff
@@ -1307,6 +1307,142 @@ public void canExtendMembershipOutOfGracePeriodAbsoluteDate() throws Exception {
+
+		Attribute triggerAttr = new Attribute(attributesManager.getAttributeDefinition(sess, AttributesManager.NS_GROUP_ATTR_DEF+":groupTrigger"));
+		triggerAttr.setValue(list);
+		attributesManager.setAttribute(sess, group3, triggerAttr);
+
+		assertTrue(groupsManager.isGroupMember(sess, group, member));
+		assertFalse(groupsManager.isDirectGroupMember(sess, group, member));
+		assertTrue(groupsManager.isGroupMember(sess, group2, member));
+		assertFalse(groupsManager.isDirectGroupMember(sess, group2, member));
+		assertTrue(groupsManager.isDirectGroupMember(sess, group3, member));
+
+		groupsManager.removeMember(sess, group3, member);
+
+		assertTrue(groupsManager.isGroupMember(sess, group, member));
+		assertTrue(groupsManager.isGroupMember(sess, group2, member));
+		assertTrue(groupsManager.isGroupMember(sess, group3, member));
+		assertFalse(groupsManager.isDirectGroupMember(sess, group, member));
+		assertFalse(groupsManager.isDirectGroupMember(sess, group2, member));
+		assertFalse(groupsManager.isDirectGroupMember(sess, group3, member));
+		assertTrue(groupsManager.isDirectGroupMember(sess, group4, member));
+	}
+
+	@Test
+	public void addMembersToGroupsFromTriggerAttributeWithGroupStructureDirectMemberOfMoreGroups() throws Exception {
+		System.out.println(CLASS_NAME + "addMembersToGroupsFromTriggerAttributeWithGroupStructureDirectMemberOfMoreGroups");
+
+		vo = setUpVo();
+
+		//set up sub groups
+		groupsManagerBl.createGroup(sess, vo, group);
+		groupsManagerBl.createGroup(sess, group, group2);
+		groupsManagerBl.createGroup(sess, group2, group3);
+		groupsManagerBl.createGroup(sess, group3, group4);
+
+		Member member = setUpMember(vo);
+		groupsManager.addMember(sess, group3, member);
+		groupsManager.addMember(sess, group4, member);
+
+		ArrayList<String> list = new ArrayList<>();
+		list.add(String.valueOf(group2.getId()));
+		ArrayList<String> list2 = new ArrayList<>();
+		list.add(String.valueOf(group4.getId()));
+
+		Attribute triggerAttr = new Attribute(attributesManager.getAttributeDefinition(sess, AttributesManager.NS_GROUP_ATTR_DEF+":groupTrigger"));
+		triggerAttr.setValue(list);
+		attributesManager.setAttribute(sess, group4, triggerAttr);
+
+		Attribute triggerAttr2 = new Attribute(attributesManager.getAttributeDefinition(sess, AttributesManager.NS_GROUP_ATTR_DEF+":groupTrigger"));
+		triggerAttr.setValue(list2);
+		attributesManager.setAttribute(sess, group3, triggerAttr2);
+
+		groupsManager.removeMember(sess, group4, member);
+		groupsManager.removeMember(sess, group3, member);
+
+		assertTrue(groupsManager.isGroupMember(sess, group, member));
+		assertTrue(groupsManager.isGroupMember(sess, group3, member));
+		assertFalse(groupsManager.isDirectGroupMember(sess, group, member));
+		assertFalse(groupsManager.isDirectGroupMember(sess, group3, member));
+		assertTrue(groupsManager.isDirectGroupMember(sess, group4, member));
+		assertTrue(groupsManager.isDirectGroupMember(sess, group2, member));
+	}
+
```

Comment:
Add another test where you will try situation like this:
- there are the group structure "A", "A:B" and the group "C"
- member M is direct in "A" and also in "A:B", he has no connection to "C"
- you will set trigger to "C" on the group "A"
- then when you remove member from "A" nothing should happen, still no connection to "C"

Reply:
Or a similiar one:
- the same group structure "A", "A:B" and "C"
- member M is direct in "A:B" and indirect in "A" (because of A:B)
- trigger is set to "C" on the group "A" again
- then when you remove member from "A:B", he should be removed automatically also from "A" and by this, he should be added directly to "C"

Reply:
If any of these test are already here, I am sorry, just don't have time to check them all.

Reply:
všechny prošly